### ISSUE TITLE
bug: Scroll Assist notWorking on SheetModal (keyboard covers input on Mobile for PWA's)

### ISSUE THREAD

**Issue by jmedellinc** [ionic-team/ionic-framework#27493]
### Prerequisites

- [X] I have read the [Contributing Guidelines](https://github.com/ionic-team/ionic-framework/blob/main/.github/CONTRIBUTING.md#creating-an-issue).
- [X] I agree to follow the [Code of Conduct](https://ionicframework.com/code-of-conduct).
- [X] I have searched for [existing issues](https://github.com/ionic-team/ionic-framework/issues) that already report this problem, without success.

### Ionic Framework Version

v7.x

### Current Behavior

The inputs when opened inside a sheetModal, get covered by the keyboard
The expected behavior is for the content to move up to make space for the keyboard so the activated input is displayed and not covered

I'm including a video of the first test using the ionic starter-app taken from my cellphone..

https://github.com/ionic-team/ionic-framework/assets/105754419/868430bb-59c3-4f44-b2a1-8ae56eccace9

Then a video with stackblitz running on the chrome inspector showing the same behavior.  I couldn't boot stackblitz on the cellphone but //fortunately// the behavior shows up on inspector. (fortunately meaning hopefully easier to debug)

https://github.com/ionic-team/ionic-framework/assets/105754419/2bffb7aa-8969-47f7-bbf9-8e23df0c4723

### Expected Behavior

Scroll Assist should auto scroll / create space to allow for the input to not being hidden behind the keyboard

### Steps to Reproduce

The stackblitz url i'm including followed the described steps but adds some additional code, to show the correct behavior on a regular modal and on a regular page

1. Create a new stackblitz starting with ion-modal controllers https://ionicframework.com/docs/api/modal#controller-modals
2. Add the breakpoints:[0.5] option to the modalController.create() command
3. Add 250px margin-top to the div containing the ion-input element in the modal-example.html
4. Verify bad behavior.

### Code Reproduction URL

https://stackblitz.com/edit/angular-cjmmkn

### Ionic Info

n/a.. running on stackblitz on latest

https://angular-cjmmkn.stackblitz.io/

### Additional Information

This was recently covered on a recent bugfix:
  https://github.com/ionic-team/ionic-framework/pull/25849
by @liamdebeasi 

But I'm guessing the sheetModal component is breaking the functionality.

There are some more scroll assist issues posted, but nothing really related to this specific case..

**Comment by liamdebeasi** [comment]
Thanks for the report. The problem here is Ionic doesn't think the input needs to scroll. As a result, `scrollAmount` is `0` and scroll assist never activates: https://github.com/ionic-team/ionic-framework/blob/70d9854d8df5259ed715e282a6ca40ca3bea6192/core/src/utils/input-shims/hacks/scroll-assist.ts#L93

However, WebKit thinks the input needs to scroll, so it will adjust the entire view anyways. In the event that scroll assist does not activate, it might be worth preventing the default browser behavior anyways, but I need to better understand the implications of doing that.

**Comment by jmedellinc** [comment]
@liamdebeasi 
Let me know if I can be of further help.

I will spend some time over the weekend to hunt for and make a quick and dirty workaround.. Would appreciate directions if you see any that would help..

Also let me know if you think I can keep an eye open about the implications you talked about..

**Comment by jmedellinc** [comment]
Hey, @liamdebeasi 

So i've got it working.. Quick and dirty is the best I could come up with.. I thought I could create a good working solution but wasn't able to.. 

I started my research on the relocateInput, jsSetFocus and the other helper methods.. Wasn't being successful.. So, looking at the big picture, I wondered whether making sure the calculation was made and finding a way to apply the padding might make things work..

Here's a main finding: Not so sure that adding a padding in the contentEl will work for modalSheets.. The sheet has the height of the full page, so there's no reason for the content to go up..

The solution i came up with is that first, when I detect that the input is inside a sheetModal (which I do by comparing contentEl.height and window.innerHeight), two things happen: 
1)  I modify the rect dimensions that are sent to getScrollData, so that the calculation of whether scroll is needed can be correctly made:
    contentElRect.height -= contentElRect.top;
    contentElRect.bottom = window.innerHeight;

2) Since the contentEl padding won't work on a sheetModal, the "padding" (which in this case is a translate - more of that below), I modify the contentEl to which the --keyboard-offset is going to be applied.. I look for parentElements until I reach an ION-MODAL.  I use the ion-modal element as contentEl, to have the variable in a usable place.

On the component, I add a class to any modal that is a sheetModal:

.modalSheetKeyboardPadding {
  transition: transform 0.5s;
  transform: translateY(calc(-1 * var(--keyboard-offset)));
}

This moves the whole modal up..
It works..

Not ideal, since there's a transparent space below the modal..
But I just couldn't find out a way to work with the translate: matrix that moves the sheetModal between the breakpoints.. I was trying to add a calc( -1 * var(--keyboard-offset) + ${currentValue}) to the Y value, but just wasn't able to figure it out.. 

I found the moveSheetToBreakpoing method and did: 
      wrapperAnimation.keyframes([
        { offset: 0, transform: `translateY(calc( -1 * var(--keyboard-offset) + ${breakpointOffset * 100}%))` },
        { offset: 1, transform: `translateY(calc( -1 * var(--keyboard-offset) + ${(1 - snapToBreakpoint) * 100}%))` },
      ]);

And it kind of works during the transitions.. But no luck when the animation ends, as expected.. just couldn't find out how the translate:matrix that places the sheetModal on the desired breakpoint comes to be..

Hopefully this might help with a proper fix?

I'd guess:
 - Proper detection of the input being in a sheetModal and calculating location based on the sheetModal location
 - Adding the keyboard-offset to the "content" part of the modal to go along with the breakpoint transform

I'd be glad to help if there's something else that may be useful..

And please let me know if you see something that I could be breaking that I should test?

Here's the patch that got it working:

````
diff --git a/node_modules/@ionic/core/dist/esm/input-shims-c7404a50.js b/node_modules/@ionic/core/dist/esm/input-shims-c7404a50.js
index 5766072..2bda6c9 100644
--- a/node_modules/@ionic/core/dist/esm/input-shims-c7404a50.js
+++ b/node_modules/@ionic/core/dist/esm/input-shims-c7404a50.js
@@ -152,10 +152,10 @@ const enableInputBlurring = () => {
 };
 
 const SCROLL_ASSIST_SPEED = 0.3;
-const getScrollData = (componentEl, contentEl, keyboardHeight) => {
+const getScrollData = (componentEl, contentElRect, keyboardHeight) => {
   var _a;
   const itemEl = (_a = componentEl.closest('ion-item,[ion-item]')) !== null && _a !== void 0 ? _a : componentEl;
-  return calcScrollData(itemEl.getBoundingClientRect(), contentEl.getBoundingClientRect(), keyboardHeight, componentEl.ownerDocument.defaultView.innerHeight // TODO(FW-2832): type
+  return calcScrollData(itemEl.getBoundingClientRect(), contentElRect, keyboardHeight, componentEl.ownerDocument.defaultView.innerHeight // TODO(FW-2832): type
   );
 };
 const calcScrollData = (inputRect, contentRect, keyboardHeight, platformHeight) => {
@@ -208,6 +208,7 @@ const PADDING_TIMER_KEY = '$ionPaddingTimer';
  */
 const setScrollPadding = (contentEl, paddingAmount, clearCallback) => {
   const timer = contentEl[PADDING_TIMER_KEY];
+  // console.log(paddingAmount,timer);
   if (timer) {
     clearTimeout(timer);
   }
@@ -220,7 +221,7 @@ const setScrollPadding = (contentEl, paddingAmount, clearCallback) => {
       if (clearCallback) {
         clearCallback();
       }
-    }, 120);
+    }, 500);
   }
 };
 /**
@@ -298,7 +299,28 @@ const jsSetFocus = async (componentEl, inputEl, contentEl, footerEl, keyboardHei
   if (!contentEl && !footerEl) {
     return;
   }
-  const scrollData = getScrollData(componentEl, (contentEl || footerEl), keyboardHeight);
+  let contentElRect=JSON.parse(JSON.stringify((contentEl || footerEl).getBoundingClientRect()));
+  // console.log('testing',contentElRect.height - window.innerHeight<20, contentElRect.top > 50);
+  if (Math.abs(contentElRect.height - window.innerHeight)<20){
+    let parent=contentEl.parentElement;
+    let i=0;
+    while(parent.nodeName != 'ION-MODAL' && i<100) {
+      parent = parent.parentElement;
+    }
+    if (parent.nodeName == 'ION-MODAL') {
+      let currentOffset=parseInt(getComputedStyle(componentEl).getPropertyValue('--keyboard-offset').replace('px',''));
+      contentElRect.height -= contentElRect.top - (currentOffset || 0);
+      contentElRect.top += (currentOffset || 0);
+      contentElRect.bottom = window.innerHeight;
+      contentEl = parent;
+      // console.log('juegue',contentElRect.height,contentElRect.bottom,contentElRect.top,window.innerHeight);
+    } else {
+      console.warn('scrollAssist sheetModal fix detonated on wrong place!');
+      debugger;
+    }
+  }
+
+  const scrollData = getScrollData(componentEl, contentElRect, keyboardHeight);
   if (contentEl && Math.abs(scrollData.scrollAmount) < 4) {
     // the text input is in a safe position that doesn't
     // require it to be scrolled into view, just set focus now

````

One interesting thing is that the 120milisecs wasn't giving the process enough time to cancel the timeout, so switching between inputs removed the keyboard-offset.. At first i realized that I needed to take into account the currentOffset for the contentElRect recalculations, but that didn't cut it.. it required the increase to 500ms for the clearTimeout.. Kind of weird..

Thanks!

**Comment by jmedellinc** [comment]
And just as I sent this, I remembered you talking about relocateInput having to disable the default behavior..

And indeed it doesn't work correctly on iOS.. Dont' have android to test it.. I just assumed that since the bug happened in chrome under iphone device simulation, fixing it there would fix it entirely.. 

The problem is dual:
* The whole app is scrolled up, so the sheetModal goes way up
* This movement happens even when the keyboard isn't shown yet (this is a problem I'm having with iOS, but well..)

Here's a video


https://github.com/ionic-team/ionic-framework/assets/105754419/0a393b67-ee23-452a-baa3-828eb3ce320e

**Comment by jmedellinc** [comment]
@liamdebeasi I will work on it this weekend again.. Do you have any further advice to guide me?
thanks!

**Comment by liamdebeasi** [comment]
Sorry for the delay. The most important component here is ensuring that the default browser behavior is prevented, so I would focus on that. In your demo video it looks like iOS is still adjusting the entire view up. Have you tried calling `relocateInput` in this if-block? https://github.com/ionic-team/ionic-framework/blob/32bc33ed28fc29ea9d6104520c370fa316606e00/core/src/utils/input-shims/hacks/scroll-assist.ts#L93-L111

**Comment by jmedellinc** [comment]
Hey @liamdebeasi 

After hammering on it, it turns out the problem I believe may be somewhere else..

The relocation of the input is working alright on sheetModals..
The problem seems to be that the input is relocated inside the sheet modal and iOS only recognizes the sheetModal space as valid, and therefore, even though it's relocated to -9999px, iOS places it at top of the sheetModal and auto-scrolls..

The conceptual solution I've come up with is that I'll clone the entire sheetModal and move the sheetModal up (and hide it with scale(0))..

The reason I'm writing is because maybe this sparks some ideas you could share with me..

Will come back later with what I find out..

BTW... This final conclusion comes after quite some debugging, testing, lot's of console logging (because I've found it hard to debug this kind of things with devTools because the debugging of scrolling behavior becomes tricky - unless I'm missing something)
The final test was that if I grow the sheetModal tall enough for the input to be in a "safe-no-auto-scroll-needed-according-to-iOS-safari", the native autoscroll stops.. 
I've ruled out the auto addition of scroll-y class to the sheetModal starting at the 0.7 breakpoint as well as other stuff I right now can't remember..

**Comment by jmedellinc** [comment]
Here again @liamdebeasi 

Ok.. So here's a "working" workaround..

The situation is two-fold:

1) Adding a padding to the bottom of ion-content doesn't work (the current mechanismo of scroll_assist).. The sheetModal.. has to be moved up..
 I implemented a transform: translateY on the ion-modal element, but has the problem that it leaves below empty space.. I think changing the breakpoint and only after that adding a padding to the content if needed would be better, but that's way above my capacity.. 
 What I did was add a class that implements the translateY using the same --keyboard-offset variable.
 The --keyboard-offset variable has to be added to the ion-modal, not to the ion-content element.  That is implemented by searching  for a modalSheet in the sheetModalDetection function in input-shims. That replaces contentEl.

2) The relocation logic was working well, but the thing is that moving the input inside the sheetModal wasn't enough, because iOS wasn't caring about the screen location of the input but rather of the maximum top of the container, which is the sheetModal, and which most of the time is below the "safe-input-location-so-dont-autoscroll" location of iOS safari..
 I'm duplicating the whole sheetModal instead of only the input, and moving it upward and to the left.  
 There's an extremely weird situation that has to do with having to have a transition effect on the modal or else it doesn't work.  I'm guessing it has to do with the sheetModal having some transitions because of it being a sheetModal.. I stumbled into the implemented solution by pure accident..

Here's the patch to input-shims:
````
diff --git a/node_modules/@ionic/core/dist/esm/input-shims-c7404a50.js b/node_modules/@ionic/core/dist/esm/input-shims-c7404a50.js
index 5766072..88a6b3d 100644
--- a/node_modules/@ionic/core/dist/esm/input-shims-c7404a50.js
+++ b/node_modules/@ionic/core/dist/esm/input-shims-c7404a50.js
@@ -8,22 +8,22 @@ import './index-e86f0117.js';
 import './index-b32cad98.js';
 
 const cloneMap = new WeakMap();
-const relocateInput = (componentEl, inputEl, shouldRelocate, inputRelativeY = 0, disabledClonedInput = false) => {
+const relocateInput = (componentEl, inputEl, shouldRelocate, inputRelativeY = 0, disabledClonedInput = false, sheetModal) => {
   if (cloneMap.has(componentEl) === shouldRelocate) {
     return;
   }
   if (shouldRelocate) {
-    addClone(componentEl, inputEl, inputRelativeY, disabledClonedInput);
+    addClone(componentEl, inputEl, inputRelativeY, disabledClonedInput, sheetModal);
   }
   else {
-    removeClone(componentEl, inputEl);
+    removeClone(componentEl, inputEl, sheetModal);
   }
 };
 // TODO(FW-2832): type
 const isFocused = (input) => {
   return input === input.getRootNode().activeElement;
 };
-const addClone = (componentEl, inputEl, inputRelativeY, disabledClonedInput = false) => {
+const addClone = (componentEl, inputEl, inputRelativeY, disabledClonedInput = false, sheetModal) => {
   // this allows for the actual input to receive the focus from
   // the user's touch event, but before it receives focus, it
   // moves the actual input to a location that will not screw
@@ -33,9 +33,15 @@ const addClone = (componentEl, inputEl, inputRelativeY, disabledClonedInput = fa
   // while the native input fakes out the browser by relocating itself
   // before it receives the actual focus event
   // We hide the focused input (with the visible caret) invisible by making it scale(0),
-  const parentEl = inputEl.parentNode;
+  const parentEl = (sheetModal || inputEl).parentNode;
+  if (sheetModal) {
+    inputEl.setAttribute('clonedInput','true')
+  }
   // DOM WRITES
-  const clonedEl = inputEl.cloneNode(false);
+  const clonedEl = (sheetModal || inputEl).cloneNode(false);
+  if (sheetModal) {
+    inputEl.removeAttribute('clonedInput');
+  }
   clonedEl.classList.add('cloned-input');
   clonedEl.tabIndex = -1;
   /**
@@ -51,23 +57,41 @@ const addClone = (componentEl, inputEl, inputRelativeY, disabledClonedInput = fa
    * if we set opacity: 1.
    */
   if (disabledClonedInput) {
-    clonedEl.disabled = true;
+    if (sheetModal) {
+      const clonedInput = clonedEl.querySelector('[clonedInput="true"]');
+      if (clnedInput) {
+        clonedInput.disabled = true;
+      } else {
+        debugger;
+        console.log('couldnt locate clonedInput');
+      }
+    } else {
+      clonedEl.disabled = true;
+    }
   }
   parentEl.appendChild(clonedEl);
   cloneMap.set(componentEl, clonedEl);
   const doc = componentEl.ownerDocument;
   const tx = doc.dir === 'rtl' ? 9999 : -9999;
   componentEl.style.pointerEvents = 'none';
-  inputEl.style.transform = `translate3d(${tx}px,${inputRelativeY}px,0) scale(0)`;
+  if (sheetModal) {
+    sheetModal.style['transition-property']='opacity';
+  }
+  (sheetModal || inputEl).style.transform = `translate3d(${tx}px,${inputRelativeY}px,0) scale(0)`;
+
 };
-const removeClone = (componentEl, inputEl) => {
+const removeClone = (componentEl, inputEl, sheetModal) => {
   const clone = cloneMap.get(componentEl);
   if (clone) {
     cloneMap.delete(componentEl);
     clone.remove();
   }
   componentEl.style.pointerEvents = '';
-  inputEl.style.transform = '';
+  (sheetModal || inputEl).style.transform = '';
+  if (sheetModal) {
+    sheetModal.style['transition-property']='';
+  }
+
 };
 /**
  * Factoring in 50px gives us some room
@@ -76,7 +100,7 @@ const removeClone = (componentEl, inputEl) => {
  */
 const SCROLL_AMOUNT_PADDING = 50;
 
-const enableHideCaretOnScroll = (componentEl, inputEl, scrollEl) => {
+const enableHideCaretOnScroll = (componentEl, inputEl, scrollEl, sheetModal) => {
   if (!scrollEl || !inputEl) {
     return () => {
       return;
@@ -84,10 +108,12 @@ const enableHideCaretOnScroll = (componentEl, inputEl, scrollEl) => {
   }
   const scrollHideCaret = (shouldHideCaret) => {
     if (isFocused(inputEl)) {
-      relocateInput(componentEl, inputEl, shouldHideCaret);
+      relocateInput(componentEl, inputEl, shouldHideCaret, null, null, sheetModal);
     }
   };
-  const onBlur = () => relocateInput(componentEl, inputEl, false);
+  const onBlur = () => {
+    relocateInput(componentEl, inputEl, false, null, null, sheetModal);
+  }
   const hideCaret = () => scrollHideCaret(true);
   const showCaret = () => scrollHideCaret(false);
   addEventListener(scrollEl, 'ionScrollStart', hideCaret);
@@ -294,10 +320,29 @@ const setManualFocus = (el) => {
   el.setAttribute(SKIP_SCROLL_ASSIST, 'true');
   el.focus();
 };
+const sheetModalDetection = (contentEl) => {
+  let parent = contentEl.parentElement;
+  while(parent?.nodeName != 'ION-MODAL' && parent?.nodeName != 'HTML') {
+    parent = parent?.parentElement;
+    if (!parent) {
+      return;
+    }
+  }
+  if (parent.nodeName == "HTML") {
+    return;
+  }
+  if (parent.classList.contains('modal-sheet')) {
+    return parent;
+  }
+}
 const jsSetFocus = async (componentEl, inputEl, contentEl, footerEl, keyboardHeight, enableScrollPadding, disableClonedInput = false) => {
   if (!contentEl && !footerEl) {
     return;
   }
+  const sheetModal = sheetModalDetection(contentEl || footerEl);
+  if (sheetModal) {
+    contentEl = sheetModal;
+  }
   const scrollData = getScrollData(componentEl, (contentEl || footerEl), keyboardHeight);
   if (contentEl && Math.abs(scrollData.scrollAmount) < 4) {
     // the text input is in a safe position that doesn't
@@ -321,7 +366,7 @@ const jsSetFocus = async (componentEl, inputEl, contentEl, footerEl, keyboardHei
   // temporarily move the focus to the focus holder so the browser
   // doesn't freak out while it's trying to get the input in place
   // at this point the native text input still does not have focus
-  relocateInput(componentEl, inputEl, true, scrollData.inputSafeY, disableClonedInput);
+  relocateInput(componentEl, inputEl, true, scrollData.inputSafeY, disableClonedInput, sheetModal);
   setManualFocus(inputEl);
   /**
    * Relocating/Focusing input causes the
@@ -354,7 +399,7 @@ const jsSetFocus = async (componentEl, inputEl, contentEl, footerEl, keyboardHei
       }
       // the scroll view is in the correct position now
       // give the native text input focus
-      relocateInput(componentEl, inputEl, false, scrollData.inputSafeY);
+      relocateInput(componentEl, inputEl, false, scrollData.inputSafeY, null, sheetModal);
       // ensure this is the focused input
       setManualFocus(inputEl);
       /**
@@ -450,7 +495,8 @@ const startInputShims = async (config, platform) => {
       return;
     }
     if (!!scrollEl && hideCaret && !hideCaretMap.has(componentEl)) {
-      const rmFn = enableHideCaretOnScroll(componentEl, inputEl, scrollEl);
+      const sheetModal = sheetModalDetection(componentEl);
+      const rmFn = enableHideCaretOnScroll(componentEl, inputEl, scrollEl, sheetModal);
       hideCaretMap.set(componentEl, rmFn);
     }
     /**
````
`_checkout the lines with transition-property_`

And the class for the sheetModal:
````
.sheetModalKeyboardPadding {
  transition: transform 0.01s;
  transform: translateY(min(0px, calc(100px + calc(-1 * (var(--keyboard-offset))))));
}
````
`_checkout the "transition: transform" line`
goes hand in hand with the _transition-property_ lines in the addClone and removeClone functions

If any of the three lines are removed, the functionality breaks

There are still some issues 
 * Changing inputs, tabbing.. I see that the blur event is working a bit erratically.. 
 * Pressing the "ready" button on the keyboard fails to close the keyboard.. It fails to really deselect the input.. You press "Ready" and the cursor disappears, but if you press any key, the key appears in the input

But the situation's much much much more manageable.. Not as much user frustration..

If you guide me in the direction for keeping investigating on these latest blur errors, I'd gladly keep hunting them..

José

**Comment by jmedellinc** [comment]
Auto height sheet modals bug out this implementation..

I believe it's somehow related to https://github.com/ionic-team/ionic-framework/issues/27846

Here's the video showing it..

https://github.com/ionic-team/ionic-framework/assets/105754419/a7b90625-a6de-40f5-82aa-8f99d7f5d671


The cloneRemoval and inputRelocation (in this case sheetModalRelocation) is not happening correctly.. The timeout has to kick in,..
Weirdly, just replacing the --height:'auto' with --height:'XXpx', solves it completely..

Since my implementation very likely won't be one you arrive to when things are fixed , i'm not opening a new issue, but if you feel that's better, I'd gladly do it..

Reporting here so it can be considered when fixing the scroll-assist code..

Let me know if I can be of further assistance @liamdebeasi 

José

**Comment by liamdebeasi** [comment]
Do you have your code changes pushed somewhere? Would be great to have a test case I can try with when fixing.
